src/df/df.rs

```diff
@@ -0,0 +1,619 @@
+#![crate_name = "uu_df"]
+
+/*
+ * This file is part of the uutils coreutils package.
+ *
+ * (c) Fangxu Hu <framlog@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE file
+ * that was distributed with this source code.
+ */
+
+extern crate libc;
+extern crate clap;
+#[macro_use] extern crate uucore;
+
+#[cfg(windows)]
+extern crate winapi;
+
+use clap::{Arg, App, ArgMatches};
+#[cfg(unix)]
+use libc::{uid_t, fsid_t};
+use std::{env, io, path, ptr, mem, slice, time};
+use std::ffi::{CStr, CString};
+use std::os::raw::{c_char, c_int};
+use std::fs::{File};
+use std::collections::HashSet;
+use std::io::{Write, BufReader, BufRead};
+use std::cell::Cell;
+use std::collections::HashMap;
+
+static VERSION: &str = env!("CARGO_PKG_VERSION");
+static ABOUT: &str = "Show information about the file system on which each FILE resides,\n\
+                      or all file systems by default.";
+
+static EXIT_OK: i32 = 0;
+static EXIT_ERR: i32 = 1;
+
+#[cfg(target_os = "linux")] static LINUX_MOUNTINFO: &str = "/proc/self/mountinfo";
+#[cfg(target_os = "linux")] static LINUX_MTAB: &str = "/etc/mtab";
+
+static OPT_ALL: &str = "all";
+static OPT_BLOCKSIZE: &str = "blocksize";
+static OPT_DIRECT: &str = "direct";
+static OPT_TOTAL: &str = "total";
+static OPT_HUMAN_READABLE: &str = "human-readable";
+static OPT_HUMAN_READABLE_2: &str = "human-readable-2";
+static OPT_INODES: &str = "inodes";
+static OPT_KILO: &str = "kilo";
+static OPT_LOCAL: &str = "local";
+static OPT_NO_SYNC: &str = "no-sync";
+static OPT_OUTPUT: &str = "output";
+static OPT_PORTABILITY: &str = "portability";
+static OPT_SYNC: &str = "sync";
+static OPT_TYPE: &str = "type";
+static OPT_PRINT_TYPE: &str = "print-type";
+static OPT_EXCLUDE_TYPE: &str = "exclude-type";
+static OPT_VERSION: &str = "version";
+
+static MOUNT_OPT_BIND: &str = "bind";
+
+/// Store names of file systems as a selector.
+/// Note: `exclude` takes priority over `include`.
+struct FsSelector {
+    include: HashSet<String>,
+    exclude: HashSet<String>,
+}
+
+struct Options {
+    show_local_fs: bool,
+    show_all_fs: bool,
+    show_listed_fs: bool,
+    show_fs_type: bool,
+    show_inode_instead: bool,
+    print_grand_total: bool,
+    block_size: i64,
+    human_readable_base: i64,
+    fs_selector: FsSelector,
+}
+
+#[derive(Debug, Clone)]
+struct MountInfo {
+    dev_id: i32,
+    dev_name: String,
+    fs_type: String,
+    mount_dir: String,
+    mount_option: String, // we only care "bind" option
+    mount_root: String,
+    remote: bool,
+    dummy: bool,
+}
+
+#[cfg(all(target_os = "macos", target_arch = "x86_64"))]
+#[repr(C)]
+#[derive(Copy, Clone)]
+#[allow(non_camel_case_types)]
+struct statfs {
```

Comment:
I believe this is available in `libc` as well.

Reply:
libc does have a `statfs` struct, but it's missing some fields which I need to use here.

Reply:
Is it missing fields for Darwin?  It looked like everything was there, but perhaps I missed something.  It does seem to be missing needed fields for FreeBSD though.

Reply:
No, that struct is complete for Darwin. And there's no need to rewrite it. sry that i didn't check that before.